File: java/saddle-points/SaddlePoints.java
import java.util.Comparator;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.HashSet;

/**
 * Saddle Points
 * https://exercism.org/tracks/java/exercises/saddle-points
 */
class Matrix {

    private final List<List<Integer>> values;

    Matrix(List<List<Integer>> values) {
        this.values = values;
    }

    Set<MatrixCoordinate> getSaddlePoints() {
        Set<MatrixCoordinate> points = new HashSet<>();
        for (int rowIdx = 0; rowIdx < values.size(); rowIdx++) {
            List<Integer> row = values.get(rowIdx);
            Integer max = Collections.max(row, Comparator.naturalOrder());
            for (int colIdx = 0; colIdx < row.size(); colIdx++) {
                Integer min = Integer.MAX_VALUE;
                for (int i = 0; i < values.size(); i++) {
                    Integer value = values.get(i).get(colIdx);
                    if (value < min) {
                        min = value;
                    }
                }

                if (max == min) {
                    points.add(new MatrixCoordinate(rowIdx + 1, colIdx + 1));
                }
            }
        }

        return points;
    }
}

// provided
class MatrixCoordinate implements Comparable<MatrixCoordinate> {
    private final int row;
    private final int col;

    MatrixCoordinate(final int row, final int col) {
        this.row = row;
        this.col = col;
    }

    @Override
    public String toString() {
        return "Row: " + row + ", Column: " + col;
    }

    // Generated equals and hashcode.

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        final MatrixCoordinate that = (MatrixCoordinate) o;

        return row == that.row && col == that.col;
    }

    @Override
    public int hashCode() {
        int result = row;
        result = 31 * result + col;
        return result;
    }

    @Override
    public int compareTo(MatrixCoordinate o) {
        int rowComparison = Integer.compare(row, o.row);
        return (rowComparison == 0) ? Integer.compare(col, o.col) : rowComparison;
    }
}

File: c/saddle-points/saddle_points.c
#include <stdlib.h>
#include "saddle_points.h"

/**
 * Saddle Points
 * https://exercism.org/tracks/c/exercises/saddle-points
 */

saddle_points_t *saddle_points(size_t rows, size_t columns, uint8_t matrix[rows][columns])
{
    saddle_points_t *points = malloc(sizeof(saddle_points_t) + rows * columns * sizeof(saddle_point_t));
    points->count = 0;

    uint8_t height_max;
    size_t row_min;
    for (size_t row = 0; row < rows; row++) {
        height_max = matrix[row][0];
        for (size_t col = 0; col < columns; col++) {
            if (matrix[row][col] > height_max) {
                height_max = matrix[row][col];
            }
        }

        for (size_t col = 0; col < columns; col++) {
            if (matrix[row][col] != height_max) {
                continue;
            }

            row_min = row;
            for (size_t r = 0; r < rows; r++) {
                if (matrix[r][col] < height_max) {
                    row_min = r;
                    break;
                }
            }

            if (row_min == row) {
                points->points[points->count].row = row + 1;
                points->points[points->count].column = col + 1;
                points->count++;
            }
        }
    }

    return points;
}

void free_saddle_points(saddle_points_t *points)
{
    free(points);
}
